src/text_metadata.rs

```diff
@@ -0,0 +1,546 @@
+    /// Encode text chunk as Vec<u8> to a `Write`
+    fn encode<W: Write>(&self, w: &mut W) -> Result<(), EncodingError>;
+}
+
+/// Struct representing a tEXt chunk
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct TEXtChunk {
+    /// Keyword field of the tEXt chunk. Needs to be between 1-79 bytes when encoded as Latin-1.
+    pub keyword: String,
+    /// Text field of tEXt chunk. Can be at most 2GB.
+    pub text: String,
+}
+
+impl TEXtChunk {
+    /// Constructs a new TEXtChunk.
+    /// Not sure whether it should take &str or String.
+    pub fn new(keyword: &str, text: &str) -> Self {
+        Self {
+            keyword: keyword.to_string(),
+            text: text.to_string(),
+        }
+    }
+
+    /// Decodes a slice of bytes to a String using Latin-1 decoding.
+    /// The decoder runs in strict mode, and any decoding errors are passed along to the caller.
+    pub(crate) fn decode(
+        keyword_slice: &[u8],
+        text_slice: &[u8],
+    ) -> Result<Self, TextDecodingError> {
+        if keyword_slice.is_empty() || keyword_slice.len() > 79 {
+            return Err(TextDecodingError::InvalidKeywordSize);
+        }
+
+        Ok(Self {
+            keyword: ISO_8859_1
+                .decode(keyword_slice, DecoderTrap::Strict)
+                .map_err(|_| TextDecodingError::Unrepresentable)?,
+            text: ISO_8859_1
+                .decode(text_slice, DecoderTrap::Strict)
+                .map_err(|_| TextDecodingError::Unrepresentable)?,
+        })
+    }
+}
+
+impl EncodableTextChunk for TEXtChunk {
+    /// Encodes TEXtChunk to a Writer. The keyword and text are separated by a byte of zeroes.
+    fn encode<W: Write>(&self, w: &mut W) -> Result<(), EncodingError> {
+        let mut data = ISO_8859_1
+            .encode(&self.keyword, EncoderTrap::Strict)
+            .map_err(|_| EncodingError::from(TextEncodingError::Unrepresentable))?;
+
+        if data.is_empty() || data.len() > 79 {
+            return Err(TextEncodingError::InvalidKeywordSize.into());
+        }
+
+        data.push(0);
+
+        ISO_8859_1
+            .encode_to(&self.text, EncoderTrap::Strict, &mut data)
+            .map_err(|_| EncodingError::from(TextEncodingError::Unrepresentable))?;
+
+        encoder::write_chunk(w, chunk::tEXt, &data)
+    }
+}
+
+/// Struct representing a zTXt chunk
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct ZTXtChunk {
+    /// Keyword field of the tEXt chunk. Needs to be between 1-79 bytes when encoded as Latin-1.
+    pub keyword: String,
+    /// Text field of zTXt chunk. It is compressed by default, but can be uncompressed if necessary.
+    pub optionally_compressed_text: OptCompressed,
+}
+
+/// Enum encoding the compressed and uncompressed states of zTXt/iTXt text field.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum OptCompressed {
+    /// Compressed version of text field. Can be at most 2GB.
+    Compressed(Vec<u8>),
+    /// Uncompressed text field.
+    Uncompressed(String),
+}
```

Comment:
That was not my original intention: the original intention was just to avoid unnecessary expensive compression/decompression when reading the text multiple times. Another option is to keep the internal text data compressed all the time, decompressing on-demand.

Reply:
It might suffice to have an opaque wrapper around this but in this manner all the variants can be constructed.

Reply:
I made the field private in the commit fb8bf59. The only way the user can set a text field is by creating a new text chunk, and they can get the text from a chunk using the get_text method.